File: backend_api/app/Models/PassengerRide.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Factories\HasFactory;

class PassengerRide extends Model
{
    //
    use HasFactory;

    protected $table = 'passenger_rides';

    /**
     * Kolom yang bisa diisi mass-assignment
     */
    protected $fillable = [
        'driver_id',
        'vehicle_type',
        'departure_terminal_id',
        'arrival_terminal_id',
        'departure_time',
        'seats_available',
        'seats_reserved',
        'price_per_seat',
        'commission_percentage',
        'ride_status',
    ];

    /**
     * ========================
     * RELATIONSHIPS
     * ========================
     */

    /**
     * Relasi ke Driver
     * Satu ride dimiliki oleh satu driver
     */
    public function driverId()
    {
        return $this->belongsTo(Driver::class, 'driver_id');
    }

    /**
     * Relasi ke Terminal Keberangkatan
     */
    public function departureTerminalId()
    {
        return $this->belongsTo(Terminal::class, 'departure_terminal_id');
    }

    /**
     * Relasi ke Terminal Tujuan
     */
    public function arrivalTerminalId()
    {
        return $this->belongsTo(Terminal::class, 'arrival_terminal_id');
    }

    /**
     * Relasi ke PassengerRideBooking
     * Satu ride bisa memiliki banyak booking
     */
    public function passengerRideBookings()
    {
        return $this->hasMany(PassengerRideBooking::class, 'passenger_ride_id');
    }

    /**
     * ========================
     * CASTS & CONSTANTS
     * ========================
     */

    protected $casts = [
        'departure_time' => 'datetime',
    ];

    public const STATUS_PENDING = 'Pending';
    public const STATUS_TRAVELING = 'Dalam Perjalanan';
    public const STATUS_DONE = 'Selesai';
    public const STATUS_CANCELED = 'Dibatalkan';

    public const VEHICLE_MOTOR = 'Motor';
    public const VEHICLE_MOBIL = 'Mobil';

    /**
     * ========================
     * HELPERS
     * ========================
     */

    /**
     * Cek apakah ride masih tersedia
     */
    public function hasAvailableSeats(): bool
    {
        return $this->seats_reserved < $this->seats_available;
    }

    /**
     * Cek apakah ride sudah selesai
     */
    public function isCompleted(): bool
    {
        return $this->ride_status === self::STATUS_DONE;
    }

    /**
     * Akses cepat untuk menghitung kursi tersisa
     */
    public function getSeatsRemainingAttribute(): int
    {
        return max(0, $this->seats_available - $this->seats_reserved);
    }

    /**
     * Hitung total pendapatan ride (belum dikurangi komisi)
     */
    public function getTotalRevenueAttribute(): int
    {
        return $this->seats_reserved * $this->price_per_seat;
    }

    /**
     * Hitung pendapatan bersih driver setelah potongan komisi
     */
    public function getDriverNetIncomeAttribute(): int
    {
        $gross = $this->getTotalRevenueAttribute();
        $commission = ($this->commission_percentage / 100) * $gross;
        return $gross - $commission;
    }
}

File: mobile_kotlin/nebeng/app/src/main/java/com/example/nebeng/feature_passenger_ride/domain/model/PassengerRide.kt
package com.example.nebeng.feature_passenger_ride.domain.model

data class PassengerRide(
    val id: Int,
//    val driverName: String,              // dari tabel drivers
    val driverId: Int,              // dari tabel drivers
    val vehicleType: String,             // dari passenger_rides
//    val departureTerminalName: String,   // dari terminals
//    val arrivalTerminalName: String,     // dari terminals
    val departureTerminalId: Int,   // dari terminals
    val arrivalTerminalId: Int,     // dari terminals
    val departureTime: String,           // dari passenger_rides
    val seatsAvailable: Int,             // dari passenger_rides
    val seatsReserved: Int,              // dari passenger_rides
    val pricePerSeat: Int,               // dari passenger_rides
    val commissionPercentage: Int,       // dari passenger_rides
    val rideStatus: String,              // dari passenger_rides
    val createdAt: String,               // dari passenger_rides
    val updatedAt: String                // dari passenger_rides
) {
    companion object {
        fun getEmpty(): PassengerRide = PassengerRide(
            id = 0,
            driverId = 0,
            departureTerminalId = 0,
            arrivalTerminalId = 0,
            vehicleType = "",
            pricePerSeat = 0,
            commissionPercentage = 0,
            seatsAvailable = 0,
            seatsReserved = 0,
            rideStatus = "",
            departureTime = "",
            createdAt = "",
            updatedAt = ""
        )
    }
}